File: sky/org.kevoree.library.sky.api/src/main/scala/org/kevoree/library/sky/api/PlanningManager.scala
package org.kevoree.library.sky.api

import org.kevoree.library.sky.api.nodeType.CloudNode
import org.kevoreeadaptation.{KevoreeAdaptationFactory, AdaptationPrimitive, ParallelStep, AdaptationModel}
import org.slf4j.{LoggerFactory, Logger}
import org.kevoree._
import org.kevoreeadaptation.impl.DefaultKevoreeAdaptationFactory
import scala.collection.JavaConversions._
import org.kevoree.kompare.{JavaSePrimitive, KevoreeKompareBean}
import org.kevoree.kompare.scheduling.SchedulingWithTopologicalOrderAlgo
import org.kevoree.framework.AbstractNodeType
import scala.collection.mutable


/**
 * User: Erwan Daubert - erwan.daubert@gmail.com
 * Date: 13/12/11
 * Time: 09:19
 *
 * @author Erwan Daubert
 * @version 1.0
 */

class PlanningManager(skyNode: AbstractNodeType) extends KevoreeKompareBean {
  private val logger: Logger = LoggerFactory.getLogger(this.getClass)

  override def compareModels(current: ContainerRoot, target: ContainerRoot, nodeName: String): AdaptationModel = {
    val factory = new DefaultKevoreeAdaptationFactory
    val adaptationModel: AdaptationModel = factory.createAdaptationModel
    //    var step: ParallelStep = factory.createParallelStep
    //    adaptationModel.setOrderedPrimitiveSet(step)
    //    if (skyNode.isHost) {
    var removeNodeType: AdaptationPrimitiveType = null
    var addNodeType: AdaptationPrimitiveType = null
    current.getAdaptationPrimitiveTypes.foreach {
      primitiveType =>
        if (primitiveType.getName == CloudNode.REMOVE_NODE) {
          removeNodeType = primitiveType
        } else if (primitiveType.getName == CloudNode.ADD_NODE) {
          addNodeType = primitiveType
        }
    }
    if (removeNodeType == null || addNodeType == null) {
      target.getAdaptationPrimitiveTypes.foreach {
        primitiveType =>
          if (primitiveType.getName == CloudNode.REMOVE_NODE) {
            removeNodeType = primitiveType
          }
          else if (primitiveType.getName == CloudNode.ADD_NODE) {
            addNodeType = primitiveType
          }
      }
    }
    if (removeNodeType == null) {
      logger.warn("there is no adaptation primitive for {}", CloudNode.REMOVE_NODE)
    }
    if (addNodeType == null) {
      logger.warn("there is no adaptation primitive for {}", CloudNode.ADD_NODE)
    }

    current.findByPath("nodes[" + skyNode.getNodeName + "]", classOf[ContainerNode]) match {
      case node: ContainerNode => {
        target.findByPath("nodes[" + skyNode.getNodeName + "]", classOf[ContainerNode]) match {
          case node1: ContainerNode => {
            node.getHosts.foreach {
              subNode =>
                node1.findByPath("hosts[" + subNode.getName + "]", classOf[ContainerNode]) match {
                  case null => {
                    logger.debug("add a {} adaptation primitive with {} as parameter", Array[String](CloudNode.REMOVE_NODE, subNode.getName))
                    val command: AdaptationPrimitive = factory.createAdaptationPrimitive
                    command.setPrimitiveType(removeNodeType)
                    command.setRef(subNode)
                    //                      val subStep: ParallelStep = factory.createParallelStep
                    //                      subStep.addAdaptations(command)
                    adaptationModel.addAdaptations(command)
                    //                      step.setNextStep(subStep)
                    //                      step = subStep
                    processStopInstance(subNode, adaptationModel, current, factory)
                  }
                  case subNode1: ContainerNode => {
                    if (subNode1.getStarted() != subNode.getStarted()) {
                      if (subNode.getStarted()) {
                        processStopInstance(subNode, adaptationModel, current, factory)
                      } else {
                        processStartInstance(subNode, adaptationModel, current, factory);
                      }
                    }
                  }
                }
            }
          }
          case null => {
            logger.debug("Unable to find the current node on the target model, We remove all the hosted nodes from the current model")
            node.getHosts.foreach {
              subNode =>
                logger.debug("add a {} adaptation primitive with {} as parameter", Array[String](CloudNode.REMOVE_NODE, subNode.getName))
                val command: AdaptationPrimitive = factory.createAdaptationPrimitive
                command.setPrimitiveType(removeNodeType)
                command.setRef(subNode)
                //                  val subStep: ParallelStep = factory.createParallelStep
                //                  subStep.addAdaptations(command)
                adaptationModel.addAdaptations(command)
                //                  step.setNextStep(subStep)
                //                  step = subStep
                processStopInstance(subNode, adaptationModel, current, factory)
            }
          }
        }
      }
      case null =>
    }

    target.findByPath("nodes[" + skyNode.getNodeName + "]", classOf[ContainerNode]) match {
      case node: ContainerNode => {
        current.findByPath("nodes[" + skyNode.getNodeName + "]", classOf[ContainerNode]) match {
          case node1: ContainerNode => {
            node.getHosts.foreach {
              subNode =>
                node1.findByPath("hosts[" + subNode.getName + "]", classOf[ContainerNode]) match {
                  case null => {
                    logger.debug("add a {} adaptation primitive with {} as parameter", Array[String](CloudNode.ADD_NODE, subNode.getName))
                    val command: AdaptationPrimitive = factory.createAdaptationPrimitive
                    command.setPrimitiveType(addNodeType)
                    command.setRef(subNode)
                    //                      val subStep: ParallelStep = factory.createParallelStep
                    //                      subStep.addAdaptations(command)
                    adaptationModel.addAdaptations(command)
                    //                      step.setNextStep(subStep)
                    //                      step = subStep
                    processStartInstance(subNode, adaptationModel, current, factory)
                  }
                  case subNode1: ContainerNode => {
                    if (subNode1.getStarted() != subNode.getStarted()) {
                      if (subNode.getStarted()) {
                        processStopInstance(subNode, adaptationModel, current, factory)
                      } else {
                        processStartInstance(subNode, adaptationModel, current, factory)
                      }
                    }
                  }
                }
            }
          }
          case null => {
            logger.debug("Unable to find the current node on the current model, We add all the hosted nodes from the target model")
            node.getHosts.foreach {
              subNode =>
                logger.debug("add a {} adaptation primitive with {} as parameter", Array[String](CloudNode.ADD_NODE, subNode.getName))
                val command: AdaptationPrimitive = factory.createAdaptationPrimitive
                command.setPrimitiveType(addNodeType)
                command.setRef(subNode)
                //                  val subStep: ParallelStep = factory.createParallelStep
                //                  subStep.addAdaptations(command)
                adaptationModel.addAdaptations(command)
                //                  step.setNextStep(subStep)
                //                  step = subStep
                processStartInstance(subNode, adaptationModel, current, factory)
            }
          }
        }
      }
      case null =>
    }
    //    }
    logger.debug("Adaptation model contain {} Host node primitives", adaptationModel.getAdaptations.size)
    //    val superModel: AdaptationModel = skyNode.superKompare(current, target)

    val superModel = super.compareModels(current, target, nodeName)

    /*if (!skyNode.isContainer && isContaining(superModel.getOrderedPrimitiveSet)) {
      throw new Exception("This node is not a container (see \"role\" attribute)")
    }*/
    adaptationModel.addAllAdaptations(superModel.getAdaptations)
    //    step.setNextStep(superModel.getOrderedPrimitiveSet)
    logger.debug("Adaptation model contain {} primitives", adaptationModel.getAdaptations.size)
    adaptationModel
  }


  private def processStopInstance(actualInstance: Instance, adaptationModel: AdaptationModel, actualRoot: ContainerRoot, adaptationModelFactory: KevoreeAdaptationFactory) {
    val ccmd2 = adaptationModelFactory.createAdaptationPrimitive()
    ccmd2.setPrimitiveType(actualRoot.findAdaptationPrimitiveTypesByID(JavaSePrimitive.instance$.getStopInstance()))
    ccmd2.setRef(actualInstance)
    adaptationModel.addAdaptations(ccmd2)
  }

  private def processStartInstance(updatedInstance: Instance, adaptationModel: AdaptationModel, updateRoot: ContainerRoot, adaptationModelFactory: KevoreeAdaptationFactory) {
    val ccmd2 = adaptationModelFactory.createAdaptationPrimitive()
    ccmd2.setPrimitiveType(updateRoot.findAdaptationPrimitiveTypesByID(JavaSePrimitive.instance$.getStartInstance()))
    ccmd2.setRef(updatedInstance)
    adaptationModel.addAdaptations(ccmd2)
  }


  override def plan(adaptationModel: AdaptationModel, p2: String): AdaptationModel = {
    if (!adaptationModel.getAdaptations.isEmpty) {

      val adaptationModelFactory = new org.kevoreeadaptation.impl.DefaultKevoreeAdaptationFactory()
      val scheduling = new SchedulingWithTopologicalOrderAlgo()
      nextStep()
      adaptationModel.setOrderedPrimitiveSet(getCurrentStep)

      // STOP child nodes
      createNextStep(adaptationModel.getAdaptations.filter {
        adapt => adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getStopInstance && adapt.getRef.isInstanceOf[ContainerNode]
      })

      // REMOVE child nodes
      createNextStep(adaptationModel.getAdaptations.filter {
        adapt => adapt.getPrimitiveType.getName == CloudNode.REMOVE_NODE
      })

      //PROCESS STOP
      scheduling.schedule(adaptationModel.getAdaptations.filter {
        adapt => adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getStopInstance && !adapt.getRef.isInstanceOf[ContainerNode]
      }, false).foreach {
        p =>
          getStep.addAdaptations(p)
          setStep(adaptationModelFactory.createParallelStep())
          getCurrentStep.setNextStep(getStep)
          setCurrentStep(getStep)
      }

      // REMOVE BINDINGS
      createNextStep(adaptationModel.getAdaptations.filter {
        adapt =>
          adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getRemoveBinding ||
            adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getRemoveFragmentBinding
      })

      createNextStep(adaptationModel.getAdaptations.filter {
        adapt => adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getRemoveInstance
      })

      createNextStep(adaptationModel.getAdaptations.filter {
        adapt => adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getRemoveType
      })

      createNextStep(adaptationModel.getAdaptations.filter {
        adapt => adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getRemoveDeployUnit
      })

      createNextStep(adaptationModel.getAdaptations.filter {
        adapt => adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getAddThirdParty
      })

      createNextStep(adaptationModel.getAdaptations.filter {
        adapt => adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getUpdateDeployUnit
      })

      createNextStep(adaptationModel.getAdaptations.filter {
        adapt => adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getAddDeployUnit
      })

      createNextStep(adaptationModel.getAdaptations.filter {
        adapt => adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getAddType
      })

      createNextStep(adaptationModel.getAdaptations.filter {
        adapt => adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getAddInstance
      })

      createNextStep(adaptationModel.getAdaptations.filter {
        adapt =>
          adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getAddBinding ||
            adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getAddFragmentBinding
      })

      createNextStep(adaptationModel.getAdaptations.filter {
        adapt => adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getUpdateDictionaryInstance
      })


      // ADD child nodes
      createNextStep(adaptationModel.getAdaptations.filter {
        adapt => adapt.getPrimitiveType.getName == CloudNode.ADD_NODE
      })

      // START child nodes
      createNextStep(adaptationModel.getAdaptations.filter {
        adapt => adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getStartInstance && adapt.getRef.isInstanceOf[ContainerNode]
      })

      var oldStep = getCurrentStep
      //PROCESS START
      scheduling.schedule(adaptationModel.getAdaptations.filter {
        adapt => adapt.getPrimitiveType.getName == JavaSePrimitive.instance$.getStartInstance && !adapt.getRef.isInstanceOf[ContainerNode]
      }, true).foreach {
        p =>
          getStep.addAdaptations(p)
          setStep(adaptationModelFactory.createParallelStep())
          getCurrentStep.setNextStep(getStep)
          oldStep = getCurrentStep
          setCurrentStep(getStep)
      }
      if (getStep.getAdaptations.isEmpty) {
        oldStep.setNextStep(null)
      }
    } else {
      adaptationModel.setOrderedPrimitiveSet(null)
    }
    clearSteps()
    adaptationModel

  }

  private def isContaining(step: ParallelStep): Boolean = {
    if (step != null) {
      // TODO must be tested
      (step.getAdaptations.exists(adaptation =>
        (adaptation.getPrimitiveType.getName == "UpdateDictionaryInstance" && !adaptation.getRef.isInstanceOf[Group])
          || (adaptation.getPrimitiveType.getName == "AddInstance" && !adaptation.getRef.isInstanceOf[Group])
          || (adaptation.getPrimitiveType.getName == "UpdateInstance" && !adaptation.getRef.isInstanceOf[Group])
          || (adaptation.getPrimitiveType.getName == "RemoveInstance" && !adaptation.getRef.isInstanceOf[Group])
          || (adaptation.getPrimitiveType.getName == "StartInstance" && !adaptation.getRef.isInstanceOf[Group])
          || (adaptation.getPrimitiveType.getName == "StopInstance" && !adaptation.getRef.isInstanceOf[Group]))
        || isContaining(step.getNextStep))
    } else {
      false
    }
  }

  protected def createNextStep(commands: java.util.List[AdaptationPrimitive]) {
    getStep.addAllAdaptations(commands)

    nextStep()
  }
}

File: sky/org.kevoree.library.sky.api/src/main/java/org/kevoree/library/sky/api/planning/PlanningManager.java
package org.kevoree.library.sky.api.planning;


import jet.runtime.typeinfo.JetValueParameter;
import org.kevoree.ContainerNode;
import org.kevoree.ContainerRoot;
import org.kevoree.Instance;
import org.kevoree.framework.AbstractNodeType;
import org.kevoree.library.defaultNodeTypes.planning.JavaSePrimitive;
import org.kevoree.library.defaultNodeTypes.planning.KevoreeKompareBean;
import org.kevoree.library.defaultNodeTypes.planning.scheduling.SchedulingWithTopologicalOrderAlgo;
import org.kevoree.library.sky.api.CloudNode;
import org.kevoree.log.Log;
import org.kevoree.modeling.api.trace.ModelAddTrace;
import org.kevoree.modeling.api.trace.ModelRemoveTrace;
import org.kevoreeadaptation.AdaptationModel;
import org.kevoreeadaptation.AdaptationPrimitive;
import org.kevoreeadaptation.ParallelStep;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * User: Erwan Daubert - erwan.daubert@gmail.com
 * Date: 13/12/11
 * Time: 09:19
 *
 * @author Erwan Daubert
 * @version 1.0
 */

public class PlanningManager extends KevoreeKompareBean {
    private AbstractNodeType skyNode;

    public PlanningManager(AbstractNodeType skyNode, Map<String, Object> registry) {
        super(registry);
        this.skyNode = skyNode;
    }

    @Override
    protected void manageModelAddTrace(@JetValueParameter(name = "trace") ModelAddTrace trace, @JetValueParameter(name = "currentNode") ContainerNode currentNode, @JetValueParameter(name = "currentModel") ContainerRoot currentModel, @JetValueParameter(name = "targetNode") ContainerNode targetNode, @JetValueParameter(name = "targetModel") ContainerRoot targetModel, @JetValueParameter(name = "adaptationModel") AdaptationModel adaptationModel) {
        if (trace.getSrcPath().equals(currentNode.path()) && trace.getRefName().equals("hosts")) {
            ContainerNode node = targetModel.findByPath(trace.getPreviousPath(), ContainerNode.class);
            addNodeInstance(node, currentModel, targetModel, adaptationModel);
        } else {
            super.manageModelAddTrace(trace, currentNode, currentModel, targetNode, targetModel, adaptationModel);
        }
    }

    @Override
    protected void manageModelRemoveTrace(@JetValueParameter(name = "trace") ModelRemoveTrace trace, @JetValueParameter(name = "currentNode") ContainerNode currentNode, @JetValueParameter(name = "currentModel") ContainerRoot currentModel, @JetValueParameter(name = "targetNode") ContainerNode targetNode, @JetValueParameter(name = "targetModel") ContainerRoot targetModel, @JetValueParameter(name = "adaptationModel") AdaptationModel adaptationModel) {
        if (trace.getSrcPath().equals(currentNode.path()) &&  trace.getRefName().equals("hosts")) {
            ContainerNode node = currentModel.findByPath(trace.getObjPath(), ContainerNode.class);
            removeNodeInstance(node, currentModel, adaptationModel);
        } else {
            super.manageModelRemoveTrace(trace, currentNode, currentModel, targetNode, targetModel, adaptationModel);
        }
    }

    private void addNodeInstance(ContainerNode subNode, ContainerRoot currentModel, ContainerRoot targetModel, AdaptationModel adaptationModel) {
        Log.debug("add a {} adaptation primitive with {} as parameter", CloudNode.ADD_NODE, subNode.getName());
        AdaptationPrimitive command = getAdaptationModelFactory().createAdaptationPrimitive();
        command.setPrimitiveType(currentModel.findAdaptationPrimitiveTypesByID(CloudNode.ADD_NODE));
        command.setRef(subNode);
        adaptationModel.addAdaptations(command);
        processStartInstance(subNode, adaptationModel, targetModel);
    }

    private void removeNodeInstance(ContainerNode subNode, ContainerRoot currentModel, AdaptationModel adaptationModel) {
        Log.debug("add a {} adaptation primitive with {} as parameter", CloudNode.REMOVE_NODE, subNode.getName());
        AdaptationPrimitive command = getAdaptationModelFactory().createAdaptationPrimitive();
        command.setPrimitiveType(currentModel.findAdaptationPrimitiveTypesByID(CloudNode.REMOVE_NODE));
        command.setRef(subNode);
        adaptationModel.addAdaptations(command);
        processStopInstance(subNode, adaptationModel, currentModel);
    }

    private void processStopInstance(Instance actualInstance, AdaptationModel adaptationModel, ContainerRoot actualRoot) {
        Log.debug("Process StopInstance on {}", actualInstance.getName());
        AdaptationPrimitive ccmd2 = getAdaptationModelFactory().createAdaptationPrimitive();
        ccmd2.setPrimitiveType(actualRoot.findAdaptationPrimitiveTypesByID(JavaSePrimitive.StopInstance));
        ccmd2.setRef(actualInstance);
        adaptationModel.addAdaptations(ccmd2);
    }

    private void processStartInstance(Instance updatedInstance, AdaptationModel adaptationModel, ContainerRoot updateRoot) {
        Log.debug("Process StartInstance on {}", updatedInstance.getName());
        AdaptationPrimitive ccmd2 = getAdaptationModelFactory().createAdaptationPrimitive();
        ccmd2.setPrimitiveType(updateRoot.findAdaptationPrimitiveTypesByID(JavaSePrimitive.StartInstance));
        ccmd2.setRef(updatedInstance);
        adaptationModel.addAdaptations(ccmd2);
    }

    @Override
    public AdaptationModel schedule(AdaptationModel adaptationModel, String nodeName) {
        if (!adaptationModel.getAdaptations().isEmpty()) {

            Log.debug("Planning adaptation and defining steps...");
            SchedulingWithTopologicalOrderAlgo scheduling = new SchedulingWithTopologicalOrderAlgo();
            nextStep();
            adaptationModel.setOrderedPrimitiveSet(getCurrentSteps());

            // STOP child nodes
            List<AdaptationPrimitive> primitives = new ArrayList<AdaptationPrimitive>();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(JavaSePrimitive.StopInstance) && primitive.getRef() instanceof ContainerNode) {
                    primitives.add(primitive);
                }
            }
            createNextStep(JavaSePrimitive.StopInstance, primitives);

            // REMOVE child nodes
            primitives.clear();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(CloudNode.REMOVE_NODE)) {
                    primitives.add(primitive);
                }
            }
            createNextStep(CloudNode.REMOVE_NODE, primitives);

            //STOP INSTANCEs (except child node)
            primitives.clear();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(JavaSePrimitive.StopInstance) && !(primitive.getRef() instanceof ContainerNode)) {
                    primitives.add(primitive);
                }
            }
            ParallelStep stepToInsert = scheduling.schedule(primitives, false);
            if (stepToInsert != null && !stepToInsert.getAdaptations().isEmpty()) {
                insertStep(stepToInsert);
            }

            // REMOVE BINDINGS
            primitives.clear();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(JavaSePrimitive.RemoveBinding) ||
                        primitive.getPrimitiveType().getName().equals(JavaSePrimitive.RemoveFragmentBinding)) {
                    primitives.add(primitive);
                }
            }
            createNextStep(JavaSePrimitive.RemoveBinding, primitives);

            // REMOVE INSTANCE
            primitives.clear();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(JavaSePrimitive.RemoveInstance)) {
                    primitives.add(primitive);
                }
            }
            createNextStep(JavaSePrimitive.RemoveInstance, primitives);

            // REMOVE TYPE
            primitives.clear();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(JavaSePrimitive.RemoveType)) {
                    primitives.add(primitive);
                }
            }
            createNextStep(JavaSePrimitive.RemoveType, primitives);

            // REMOVE TYPE
            primitives.clear();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(JavaSePrimitive.RemoveDeployUnit)) {
                    primitives.add(primitive);
                }
            }
            createNextStep(JavaSePrimitive.RemoveDeployUnit, primitives);

            // REMOVE THIRD PARTY
            primitives.clear();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(JavaSePrimitive.AddThirdParty)) {
                    primitives.add(primitive);
                }
            }
            createNextStep(JavaSePrimitive.AddThirdParty, primitives);

            // UPDATE DEPLOYUNITs
            primitives.clear();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(JavaSePrimitive.UpdateDeployUnit)) {
                    primitives.add(primitive);
                }
            }
            createNextStep(JavaSePrimitive.UpdateDeployUnit, primitives);

            // ADD DEPLOYUNITs
            primitives.clear();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(JavaSePrimitive.AddDeployUnit)) {
                    primitives.add(primitive);
                }
            }
            createNextStep(JavaSePrimitive.AddDeployUnit, primitives);

            // ADD TYPEs
            primitives.clear();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(JavaSePrimitive.AddType)) {
                    primitives.add(primitive);
                }
            }
            createNextStep(JavaSePrimitive.AddType, primitives);

            // ADD INSTANCEs
            primitives.clear();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(JavaSePrimitive.AddInstance)) {
                    primitives.add(primitive);
                }
            }
            createNextStep(JavaSePrimitive.AddInstance, primitives);

            // ADD BINDINGs
            primitives.clear();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(JavaSePrimitive.AddBinding) ||
                        primitive.getPrimitiveType().getName().equals(JavaSePrimitive.AddFragmentBinding)) {
                    primitives.add(primitive);
                }
            }
            createNextStep(JavaSePrimitive.AddBinding, primitives);

            // UPDATE DICTIONARY
            primitives.clear();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(JavaSePrimitive.UpdateDictionaryInstance)) {
                    primitives.add(primitive);
                }
            }
            createNextStep(JavaSePrimitive.UpdateDictionaryInstance, primitives);


            // ADD child nodes
            primitives.clear();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(CloudNode.ADD_NODE)) {
                    primitives.add(primitive);
                }
            }
            createNextStep(CloudNode.ADD_NODE, primitives);

            // START child nodes
            primitives.clear();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(JavaSePrimitive.StartInstance) && primitive.getRef() instanceof ContainerNode) {
                    primitives.add(primitive);
                }
            }
            createNextStep(JavaSePrimitive.StartInstance, primitives);

            // START INSTANCEs (except child nodes)
            primitives.clear();
            for (AdaptationPrimitive primitive : adaptationModel.getAdaptations()) {
                if (primitive.getPrimitiveType().getName().equals(JavaSePrimitive.StartInstance) && !(primitive.getRef() instanceof ContainerNode)) {
                    primitives.add(primitive);
                }
            }
            stepToInsert = scheduling.schedule(primitives, true);
            if (stepToInsert != null && !stepToInsert.getAdaptations().isEmpty()) {
                insertStep(stepToInsert);
            }
        } else {
            adaptationModel.setOrderedPrimitiveSet(null);
        }
        clearSteps();
        return adaptationModel;

    }
}